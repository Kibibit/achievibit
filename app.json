{
  "name": "achievibit",
  "description": "Github Gamification - Achievements system as a GitHub WebHook. http://achievibit.kibibit.io/",
  "scripts": {
  },
  "env": {
  },
  "formation": {
  },
  "addons": [

  ],
  "buildpacks": [
    {
      "url": "heroku/nodejs"
    }
  ]
}
