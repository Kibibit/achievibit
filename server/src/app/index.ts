/**
 * @file Automatically generated by barrelsby.
 */

export * from './app.controller';
export * from './app.module';
