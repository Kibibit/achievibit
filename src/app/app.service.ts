import { Injectable } from '@nestjs/common';

import { ConfigService } from '@kb-config';
import { PackageDetailsDto } from '@kb-models';

@Injectable()
export class AppService {
  constructor(private configService: ConfigService) { }
  getPackageDetails(): PackageDetailsDto {

    return this.configService.packageDetails;
  }
}
