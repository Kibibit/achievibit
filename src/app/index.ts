/**
 * @file Automatically generated by barrelsby.
 */

export * from './app.controller';
export * from './app.module';
export * from './app.service';
