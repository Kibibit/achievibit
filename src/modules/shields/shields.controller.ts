import { Controller, Get, Header, Redirect, HttpStatus } from '@nestjs/common';
import { ApiOperation, ApiTags, ApiResponse } from '@nestjs/swagger';

import { ShieldsService } from './shields.service';

@Controller('api/shields')
@ApiTags('Shields')
export class ShieldsController {
  constructor(private readonly shieldsService: ShieldsService) { }

  @Get()
  @Redirect('api/shields/achievements', HttpStatus.MOVED_PERMANENTLY)
  @ApiOperation({ summary: 'Redirects to api/shields/achievements' })
  @ApiResponse({ status: HttpStatus.MOVED_PERMANENTLY })
  redirectToGetAchievementsShield() {

  }

  @Get('achievements')
  @Header('Content-Type', 'image/svg+xml;charset=utf-8')
  @Header('Pragma-directive', 'no-cache')
  @Header('Cache-directive', 'no-cache')
  @Header('Pragma', 'no-cache')
  @Header('Expires', '0')
  @Header('Cache-Control', 'no-cache, no-store, must-revalidate')
  @ApiOperation({ summary: 'Returns achievements shield' })
  async getAchievementsShield(): Promise<string> {
    return await this.shieldsService.createAchievementsShield();
  }
}
