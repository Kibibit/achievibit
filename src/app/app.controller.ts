import {
  Body,
  ClassSerializerInterceptor,
  Controller,
  Get,
  Headers,
  HttpStatus,
  Post,
  Render,
  UseInterceptors,
} from '@nestjs/common';
import { ApiOperation, ApiResponse } from '@nestjs/swagger';
import { groupBy } from 'lodash';

import { PackageDetailsDto } from '@kb-models';
import { GithubEventManagerService, ReposService, UsersService } from '@kb-modules';

import { AppService } from './app.service';

@Controller()
export class AppController {
  constructor(
    private readonly appService: AppService,
    private readonly reposService: ReposService,
    private readonly usersService: UsersService,
    private readonly githubEventManagerService: GithubEventManagerService
  ) { }

  @Get()
  @ApiOperation({ summary: `Return achievibit's homepage` })
  @Render('index.njk')
  async getHomepage(): Promise<any> {
    const repos = await this.reposService.findAll();
    const allUsers = await this.usersService.findAll();
    const groupedUsers = groupBy(allUsers, 'organization');
    const users = groupedUsers.undefined;
    const organizations = groupedUsers.true;

    return { repos, users, organizations };
  }

  @Get('api')
  @ApiOperation({ summary: `API Hello endpoint. Returns achievebit's package information` })
  @ApiResponse({
    status: HttpStatus.OK,
    description: `achievibit's package information`,
    type: PackageDetailsDto
  })
  @UseInterceptors(ClassSerializerInterceptor)
  async getAchievibitPackageInfo(): Promise<PackageDetailsDto> {
    return await this.appService.getPackageDetails();
  }

  @Post()
  @ApiOperation({ summary: `Recieve GitHub Webhooks` })
  async recieveGitHubWebhooks(@Headers('x-github-event') githubEvent: string, @Body() eventBody: any) {
    await this.githubEventManagerService.postFromWebhook(githubEvent, eventBody);

    return;
  }
}
