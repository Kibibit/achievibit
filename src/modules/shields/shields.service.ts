import { Injectable, Logger } from '@nestjs/common';
import { BadgeFactory } from 'gh-badges';
import { keys } from 'lodash';
import requireAll from 'require-all';

import { AppService } from '../../app.service';

@Injectable()
export class ShieldsService {
  private logger: Logger = new Logger('ShieldsService');
  private bf: BadgeFactory;

  achievements: any;

  constructor(private readonly appService: AppService) {
    this.achievements = requireAll({
      dirname: appService.appRoot + '/achievements',
      filter: /(.+achievement)\.js$/,
      excludeDirs: /^\.(git|svn)$/,
      recursive: true
    });

    this.logger.log('These achievements were registered:');
    this.logger.log(this.achievements);

    this.bf = new BadgeFactory();
  }

  async createAchievementsShield(): Promise<string> {
    const format = {
      format: 'svg',
      text: [ 'achievements', keys(this.achievements).length ],
      labelColor: '#894597',
      color: '#5d5d5d',
      template: 'for-the-badge',
      logo: [
        'data:image/png;base64,iVBORw0KGgoAAAA',
        'NSUhEUgAAACAAAAAgCAYAAABzenr0AAAABmJL',
        'R0QA/wD/AP+gvaeTAAAA/0lEQVRYhe3WMU7DM',
        'BjFcadqh0qdWWBl7QU4Ss/AjsREF8RdOhYO0E',
        'qoN2DhFIgBOvBjIIMVxSFyUiEhP8lD7C/v/T9',
        '7sEMoKkoIe+Npn8qpOgCM2VBVVa1ZkzFDcjQd',
        'apDqLIR+u/jnO1AACkABKABdAO9DjHEWfb7lA',
        'LwOAQghXPXx6gJ4zE3GJIRwE0095Zhc4PO3iz',
        '7x7zoq+cB5bifr9tg0AK7xFZXcZYXXZjNs+wB',
        'giofG8hazbIDaeI5dFwAu8dxY2mE+KDyCWGCT',
        'YLj3c86xNliMEh5BVLjFseNEjnVN8pU0BsgSh',
        '5bwA5YnC25AVFjhpR6rk3Zd9K/1Dcae2pUn6m',
        'qiAAAAAElFTkSuQmCC'
      ].join('')
    };

    return this.bf.create(format);
  }
}
