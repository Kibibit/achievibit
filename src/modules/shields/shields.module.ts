import { Module } from '@nestjs/common';

import { ConfigModule } from '@kb-config';

import { ShieldsController } from './shields.controller';
import { ShieldsService } from './shields.service';

@Module({
  imports: [ ConfigModule ],
  providers: [ ShieldsService ],
  controllers: [ ShieldsController ]
})
export class ShieldsModule { }
