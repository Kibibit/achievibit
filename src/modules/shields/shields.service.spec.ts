import { Test, TestingModule } from '@nestjs/testing';
import { JSDOM } from 'jsdom';
import { keys, noop } from 'lodash';
import requireAll from 'require-all';

import { DtoMockGenerator } from '@kb-dev-tools';

import { AppService } from '../../app.service';
import { ShieldsService } from './shields.service';

jest.mock('require-all');

const mocked = requireAll as jest.Mocked<typeof requireAll>;

const mockedAchievements = DtoMockGenerator.achievementScripts();

mocked.mockReturnValue(mockedAchievements);

describe('ShieldsService', () => {
  let service: ShieldsService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        { provide: AppService, useValue: { getPackageDetails: noop } },
        ShieldsService
      ]
    }).compile();

    service = module.get<ShieldsService>(ShieldsService);
  });

  it('should be defined', async () => {
    expect(service).toBeDefined();
  });

  it('should create an all achievements shield with number of defined achievements', async () => {
    const twoAchievements = DtoMockGenerator.achievementScripts(2);
    service.achievements = twoAchievements;
    const svgAchievementsShield = await service.createAchievementsShield();

    const dom = new JSDOM(svgAchievementsShield);

    const allSvgTextElements: SVGTextElement[] = Array.from(dom.window.document.querySelectorAll('text'));
    const allSvgTextContent: string[] = allSvgTextElements.map((svg) => svg.textContent);

    const numberOfAchievements = keys(twoAchievements).length.toString();

    expect(allSvgTextContent).toContain(numberOfAchievements);
    // compare the svg output for changes
    expect(svgAchievementsShield).toMatchSnapshot();
  });
});
