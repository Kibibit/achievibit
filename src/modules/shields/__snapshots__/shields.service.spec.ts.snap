// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShieldsService should create an all achievements shield with number of defined achievements 1`] = `"<svg xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" width=\\"161\\" height=\\"28\\"><g shape-rendering=\\"crispEdges\\"><path fill=\\"#894597\\" d=\\"M0 0h132v28H0z\\"/><path fill=\\"#5d5d5d\\" d=\\"M132 0h29v28H132z\\"/></g><g fill=\\"#fff\\" text-anchor=\\"middle\\" font-family=\\"DejaVu Sans,Verdana,Geneva,sans-serif\\" font-size=\\"100\\"><image x=\\"9\\" y=\\"7\\" width=\\"14\\" height=\\"14\\" xlink:href=\\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAABmJLR0QA/wD/AP+gvaeTAAAA/0lEQVRYhe3WMU7DMBjFcadqh0qdWWBl7QU4Ss/AjsREF8RdOhYO0EqoN2DhFIgBOvBjIIMVxSFyUiEhP8lD7C/v/T97sEMoKkoIe+Npn8qpOgCM2VBVVa1ZkzFDcjQdapDqLIR+u/jnO1AACkABKABdAO9DjHEWfb7lALwOAQghXPXx6gJ4zE3GJIRwE0095Zhc4PO3iz7x7zoq+cB5bifr9tg0AK7xFZXcZYXXZjNs+wBgiofG8hazbIDaeI5dFwAu8dxY2mE+KDyCWGCTYLj3c86xNliMEh5BVLjFseNEjnVN8pU0BsgSh5bwA5YnC25AVFjhpR6rk3Zd9K/1Dcae2pUn6mqiAAAAAElFTkSuQmCC\\"/> <text x=\\"745\\" y=\\"175\\" transform=\\"scale(.1)\\" textLength=\\"910\\">ACHIEVEMENTS</text><text x=\\"1465\\" y=\\"175\\" font-weight=\\"bold\\" transform=\\"scale(.1)\\" textLength=\\"50\\">2</text></g> </svg>"`;
